sphinx/source/conf.py

```diff
@@ -275,3 +270,76 @@
     'pandas': ('http://pandas.pydata.org/pandas-docs/stable/', None),
     'numpy': ('https://docs.scipy.org/doc/numpy/', None)
 }
+
+
+import os
+import subprocess
+
+# Do not use Bokeh CDNs to serve JS files when building latest version
+# of docs (and/or unreleased versions). We want to serve the compiled
+# JS files from Read the Docs servers itself to match the
+# documentation with the JS version. Although, when using a tagged
+# release, we want to use Bokeh CDN.
+def _get_current_commit():
+    result = subprocess.run(
+        ['git', 'rev-parse', '--verify', 'HEAD'],
+        stdout=subprocess.PIPE,
+    )
+    return result.stdout.decode('utf8').strip()
+
+def is_tagged_version():
+    commit = _get_current_commit()
+    print('Bokeh commit: {commit}'.format(commit=commit))
+    result = subprocess.run(
+        ['git', 'tag', '--points-at', commit],
+        stdout=subprocess.PIPE,
+    )
+    tagged = result.stdout.decode('utf8').strip()
+    print('Bokeh tagged version: {tagged}'.format(tagged=tagged))
+    return bool(tagged)
+
+if os.environ.get('READTHEDOCS') == 'True':
+    readthedocs_version = os.environ.get('READTHEDOCS_VERSION')
+    if is_tagged_version():
+        # Use Bokeh Resources from CDN
+        os.environ['BOKEH_DOCS_CDN'] = readthedocs_version
+    else:
+        # Use Bokeh Resources from Read the Docs
+        os.environ['BOKEH_DOCS_CDN'] = 'test:{readthedocs_version}'.format(
+            readthedocs_version=readthedocs_version,
+        )
+
+    bokeh_docs_cdn = os.environ.get('BOKEH_DOCS_CDN')
+    print('BOKEH_DOCS_CDN: {bokeh_docs_cdn}'.format(
+        bokeh_docs_cdn=bokeh_docs_cdn,
+    ))
+
+    # Set the proper ``BOKEH_DOCS_VERSION`` based on the version being
+    # built on Read the Docs
+    os.environ['BOKEH_DOCS_VERSION'] = readthedocs_version
+    print('BOKEH_DOCS_VERSION: {readthedocs_version}'.format(
+        readthedocs_version=readthedocs_version,
+    ))
+
+    build_js = os.environ.get('BOKEH_BUILD_JS')
+    print('BOKEH_BUILD_JS: {build_js}'.format(build_js=build_js))
+
+    # Download all the sample data to plot
+    import bokeh
+    bokeh.sampledata.download(progress=False)
```

Comment:
i don't think this should happen in `conf.py` In general I think `conf.py` should try to avoid executing code much beyond simple assignments. But this this is also wasteful in the local usage scenario, both of our CDN bandwidth as well as of users time having to re-download data even time they do a build. The docs build should assume sample data is installed (and it should be done as part of the build process up front)

Reply:
This is under a `if` clause, and will be executed _only_ when the build is ran under Read the Docs (`READTHEDOCS` environment variable will be `True`).

See https://docs.readthedocs.io/en/stable/builds.html#build-environment for more details on environment variables defined by Read the Docs.

Reply:
I missed above that this was in a conditioned branch. I suppose this OK tho it might be nice to move some of this out to a support module. I like to keep `setup.py` and `conf.py` etc as short and simple as possible